docs/learn.md

```diff
@@ -721,20 +723,1250 @@ template](https://github.com/melange-re/melange-opam-template).
+```
+
+This generates:
+
+```js
+process.on("exit", function (exitCode) {
+  console.log("error code: " + exitCode.toString());
+});
+```
+
+The `bs.as "exit"` and the wildcard `_` pattern together will tell Melange to
+compile the first argument of the JavaScript function to the string `"exit"`.
+
+You can also use any JSON literal by passing a quoted string to `bs.as`: `bs.as
+{json|true|json}` or `bs.as {json|{"name": "John"}|json}`.
+
+#### Binding to callbacks
+
+In OCaml, all functions have arity 1. This means that if you define a function
+like this:
+
+```ocaml
+let add x y = x + y
+```
+
+Its type will be `int -> int -> int`. This means that one can partially apply
+`add` by calling `add 1`, which will return another function expecting the
+second argument of the addition. This kind of functions are called "curried"
+functions, more information about currying in OCaml can be found in [this
+chapter](https://cs3110.github.io/textbook/chapters/hop/currying.html) of the
+"OCaml Programming: Correct + Efficient + Beautiful" book.
+
+Currying is problematic when binding to JavaScript, let’s say we have a function
+like this:
+
+```javascript
+function map (a, b, f){
+  var i = Math.min(a.length, b.length);
+  var c = new Array(i);
+  for(var j = 0; j < i; ++j){
+    c[j] = f(a[i],b[i])
+  }
+  return c ;
+}
+```
+
+A naive external function declaration could be as below:
+
+```ocaml
+external map : 'a array -> 'b array -> ('a -> 'b -> 'c) -> 'c array = "map"
+  [@@bs.val]
+```
+
+Unfortunately, this is not completely correct. The issue is in the callback
+function, with type `'a → 'b → 'c`. This means that `map` will expect a function
+like `add` described above. But as we said, in OCaml, having two arguments means
+just to have two functions that take one argument.
+
+Let’s rewrite `add` to make the problem a bit more clear:
+
+```ocaml
+let add x = let partial y = x + y in partial
+```
+
+This will be compiled to:
+
+```javascript
+function add(x) {
+  return (function (y) {
+    return x + y | 0;
+  });
+}
+```
+
+Now if we ever used our external function `map` with our `add` function by
+calling `map arr1 arr2 add` it would not work as expected. JavaScript function
+application does not work the same as in OCaml, so the function call in the
+`map` implementation, `f(a[i],b[i])`, would be applied over the outer JavaScript
+function `add`, which only takes one argument `x`, and `b[i]` would be just
+discarded. The value returned from the operation would not be the addition of
+the two numbers, but rather the inner anonymous callback.
+
+To solve this mismatch between OCaml and JavaScript functions and their
+application, Melange provides a special attribute `@bs` that can be used with
+external functions.
+
+In the example above:
+
+```ocaml
+external map : 'a array -> 'b array -> (('a -> 'b -> 'c)[@bs]) -> 'c array
+  = "map"
+  [@@bs.val]
+```
+
+Here `('a → 'b → 'c [@bs])` will be interpreted as having arity 2, in general,
```

Comment:
what's this `→` character? can we use `->` instead?

Reply:
Removed in https://github.com/melange-re/documentation-site/pull/12/commits/ce220da061bc231ee5c1cb4f6ab96a2ce0167746 (also replaced the `…​` weird char with `...`)